sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -146,8 +146,12 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
   }
 
   def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
-    case PhysicalOperation(projects, filters,
+    case ScanOperation(projects, allFilters,
       l @ LogicalRelation(fsRelation: HadoopFsRelation, _, table, _)) =>
+      // We can only push down the bottom-most filter to the relation, as `ScanOperation` decided to
+      // not merge these filters and we need to keep their evaluation order.
+      val filters = allFilters.lastOption.getOrElse(Nil)
```

Comment:
So for filter pushdown, we will use the last filter. For schema pruning, we will use all the filters. 
I wonder if we should return both `allFilters` and `pushdownFilters` to make the syntax clear.

Reply:
This is the only place that does column pruning and filter pushdown together. Making `ScanOperation` return both `allFilters` and `pushdownFilters` will make the code more verbose in other places.